### ISSUE TITLE
Allow different stretching algorithms to be used during export

### ISSUE THREAD

**Issue by petersampsonaudacity** [audacity/audacity#5603]
TL;DR of below discussion:

* the SP stretcher correlates both sides of the stereo signal to make sure transients don't cancel out
* In cases of highly uncorrelated stereo, the signal may appear to wander from one channel to the other
* Other time stretching algorithms may not have this issue. However, SBSMS in particular is orders of magnitude too slow for realtime and thus would need to be selectable in the export window instead.

### Bug description
Clip Speed Rendering of a stretched clip destroys a stereo track's stereo image.  

This happens either if you speed render a stretched clip explicitly with the Render Clip Stretching context menu command or if you apply an effect to a stretched clip (as that will auto render prior to applying the effect).

This effectively loses a user's data as it is not readily possible to restore the stereo image later

### Steps to reproduce
1. import a known stereo track with clear, extreme, stereo imaging (I use track 1 of Sgt Pepper)
2. Time Stretch the clip shorter to 125% speed
3. Observe: Stereo image remains visually and aurally intact - **_UPDATE looks Ok visually BUT the stereo image is already corrupted/destroyed at this stage_**
4. use **Render Clip Stretching** from the clip-handle context menu (or from the Edit > Audio Clips submenu)
5. Observe: the waveforms in both channels are largely similar - the stereo image looks to be destroyed
6. Press Play
7. Observe the stereo sound stage has disappeared with the audio sounding centered _(destroying George Martin's careful work in the Abbey Road studios)_

### Expected behavior
The stereo image should not be altered by rendering a stretched clip

### Actual behavior
**Step 1**
![image](https://github.com/audacity/audacity/assets/50205984/65cff9f0-d6e5-49d2-91bd-59993d731461)

**Step 3**
![image](https://github.com/audacity/audacity/assets/50205984/59948a17-b69a-41b5-a24a-b6a9be60ec5f)

**Step 5**
![image](https://github.com/audacity/audacity/assets/50205984/7b0b2815-1e2e-4fdf-8b81-3a2fe54fd3fd)

### Audacity Version
3.4.1 and latest 3.4.2 beta (in fact all 3.4)

### Operating system
Windows 10 - but assume all OS

### Additional context
Originally discovered in #5588 @saintmatthieu  suggested that this should be an enhancement but I _**strongly**_ believe that is is a bug as is loses users' stereo data which will be hard or impossible (in some cases) to recover later.

**Comment by petersampsonaudacity** [comment]
This  perhaps implies, that each of the two stereo channels should be Clip Speed rendered separately.

I certainly get a better result with Sgt Pepper if I split the stereo track to mono after the stretch to 125% and then Clip Speed Render each mono track separately and the rejoin them to be a stereo track:
![image](https://github.com/audacity/audacity/assets/50205984/65af4ea8-4920-451e-bc29-eabe415ac3bb)
Proper stereo with this

**Comment by petersampsonaudacity** [comment]
And here's an extreme example of stereo image destruction _(more like mangling - chew it up and spit it out!)_
1. generate a chirp
2. generate a DTMF tone on a new track
3. make the mono pair a stereo track
4. Clip-stretch to 125% speed
5. Effect > Amplify
6. Observe

![image](https://github.com/audacity/audacity/assets/50205984/6e6ac5b7-3dbc-44d8-8e52-859f75cf5b4c)

_It's the render that does the damage not the Amplify - if you Clip Speed render at step 5 you get a similar result, just with lower amplitude._

@saintmatthieu  Does this imply, perhaps, that each of the two stereo channels should be Clip Speed rendered separately??

**Comment by saintmatthieu** [comment]
#### A technical explanation

The algorithm resynthesizes the audio at lower or higher speed unwinding the phases of the input's frequency components (partials). Because a component is generally represented by more than 1 FFT bin, and all bins contributing to a particular component must have consistent phases, the bins are grouped component-wise. Say I have a signal consisting of two sinusoids, then maybe bins 10 to 15 will contribute to component 1, and bins 100 to 105, to component 2.

For the grouping, a magnitude peak-finding and -tracking algorithm is used. For stereo signals, treating left and right channels independently make it possible that this algorithm takes different decisions, leading to inconsistencies such as this phase inversion (from project attached to #5390):
![image](https://github.com/audacity/audacity/assets/22740106/fe19ca34-766f-48b0-81dd-758fd2eecd68)

To solve this, the algorithm assumes that left and right channels really are the two, almost identical sides of a stereo image. It transforms left and right to mid- and side channels, and, only uses the mid-channel for the peak tracking. The phase unwinding is then consistent between the channels.

When left and right channels are different, though, the averaging makes little sense, and so do the results.

#### TL;DR

I recognize that this is an imperfection of our implementation. Maybe we can find a metric of left-right similarity that correlates dual-mono-processing (as @petersampsonaudacity you suggest doing) with better quality, and then use it to decide which approach to take.

I don't think that something is broken, though, and would stick with "enhancement request". (The distinction in this case may not be so important, though.)

**Comment by petersampsonaudacity** [comment]
I just take a simple-minded end-user approach to this.  

I can already forsee [posts on the Forum and Facebook along the lines:
 "Why has Audacity turned my stereo into mono ?

To which I reply is:
"Sorry, that's just the way the Time Stretch Rendering works.

End users don't care about the details of the algorithm - what they do care about is the integrity and quality of their Audio in Audacity.

**Comment by crsib** [comment]
I can hardly imagine the use case when one needs to time stretch a stereo track with a wide stereo field. In any case, there are alternatives **in** Audacity if the user is unhappy with the result this specific algorithm provides. 

Time-stretching algorithms are not perfect, nor is virtually any audio processing algorithm. That's why there are thousands of effects out there. Waves alone has 22 equalizers, 19 reverbs, and 16 limiters. 

So yes,

> Sorry, that's just the way the Time Stretch Rendering works ...

**Comment by SteveDaulton** [comment]
One possible workaround might be to give the user an option to choose whether to use the "real-time" algorithm or sbsms when rendering the stretched audio. The "real-time" algorithm works pretty well with music,  so long as there's not too much stereo spread, but sbsms is better on some kinds of sounds (particularly synth sounds) and handles wide stereo better.

**Comment by saintmatthieu** [comment]
@petersampsonaudacity 

> Steps to reproduce
> 1. import a known stereo track with clear, extreme, stereo imaging (I use track 1 of Sgt Pepper)
> 2. Time Stretch the clip shorter to 125% speed
> 3. Observe: Stereo image remains visually and aurally intact

Just had a conversation with @Paul-Licameli , who made me realize the "_aurally_ intact" part of your comment - sorry I overlooked that.
I assume you mean that if you use normal playback, the stereo image remains intact. Just did the chirp-DTMF experiment. Left and right channels leak into each other during playback whether the stretch was rendered offline before playback or not. You may want to double-check this.

**Comment by petersampsonaudacity** [comment]
> I assume you mean that if you use normal playback, the stereo image remains intact. ... You may want to double-check this.

@saintmatthieu 

Yes you're right - now I listen more carefully **(difficult as the headphone output jack on my laptop no loner works so I have to rely on onboard speakers)** the stereo image sounds aurally destroyed - but still looks like the original waveform.

I see no advice to the users regarding this on the Muse Audacity support site:  https://support.audacityteam.org/

And I have certainly not written about this in the manual - should I ?

**Comment by petersampsonaudacity** [comment]
In fact this page on the support site has no mention of time stretching for clips
https://support.audacityteam.org/basics/audacity-editing

**Comment by LWinterberg** [comment]
no, but https://support.audacityteam.org/audio-editing/speeding-up-and-slowing-down-audio has.

**Comment by petersampsonaudacity** [comment]
I have added a warning in the Manual:

![image](https://github.com/audacity/audacity/assets/50205984/8f876b2d-29a3-4504-af15-42b87b15d2a1)

**Comment by crsib** [comment]
But this is not quite true. It only affects extreme cases, aint it?

**Comment by petersampsonaudacity** [comment]
Hard to tell Dmitry.  You on;y notice if you have tho odd extreme stereo separation (like that which George Martin gave to the Beatles records in the late sixties - OKish on speakers, but most odd on cans).

If the stereo spread is not wide then it becomes harder to tell, harder to hear if it's changed/damaged.

**Comment by crsib** [comment]
In our testing, running time stretching on a stereo track generally gives better results than running in a dual mono. The difference was big enough for us to spot a bug once: #5390.

So, I strongly disagree with your assertion in the manual.

**Comment by petersampsonaudacity** [comment]
> So, I strongly disagree with your assertion in the manual.

@crsib 

So do you want me to redact that  ?

Or tone it down to _"may alter the stereo image ..."_

Peter

**Comment by petersampsonaudacity** [comment]
@saintmatthieu @crsib 

Are you happier with this revised version
![image](https://github.com/audacity/audacity/assets/50205984/9b3b7da3-f5b3-488d-a678-982e3ddf8abe)

**Comment by saintmatthieu** [comment]
Yes. How about describing the workaround you found of splitting to mono first ? Maybe mentioning alternative time-stretch processors such as SBSMS, too. That could help some users.

**Comment by SteveDaulton** [comment]
> In our testing, running time stretching on a stereo track generally gives better results than running in a dual mono.

I'd agree with that in terms of overall  sound quality, but even with  just 90% stretch (-10% in Change Tempo) I can clearly hear that the stereo field has been mangled, (which does not happen with sbsms). I don't even need to use exceptionally wide stereo to notice the difference - orchestral  music through decent headphones reveals the mangled stereo field.

**Comment by petersampsonaudacity** [comment]
> Yes. How about describing the workaround you found of splitting to mono first ?

OK - so this is what I did:
![image](https://github.com/audacity/audacity/assets/50205984/94a25113-0b2b-4173-83b0-b03156816ebb)

**Comment by petersampsonaudacity** [comment]
> but even with just 90% stretch (-10% in Change Tempo) I can clearly hear that the stereo field has been mangled

I can hear that the stereo gets changed with even a 1% change to 101% speed

the track I have been testing with today is one of my favourite jazz tracks Sonny Rollins' Without A Song from his The Bridge album.  The engineers created a wonderfully realistic spread stereo sound stage (better then the somewhat artificial stuff that George Martin did for the earlier Beatles works).

I can supply a copy of that track if anybody in Muse would like it for testing.

**Comment by SteveDaulton** [comment]
An easier workaround when the stereo imaging is important is to use Audacity's old  Change Tempo effect (high quality).

(Both workarounds are "destructive" since joining tracks to stereo causes the stretched tracks to be rendered)

**Comment by saintmatthieu** [comment]
Good morning,
I still disagree with the title "Time-stretching is best performed on mono clips." While you'll probably find examples where dual-mono processing yields better results, you'll also find others where it's the other way round, hence the generalization is in my opinion wrong. Why not "Time-stretching may affect stereo image" ?

**Comment by saintmatthieu** [comment]
> An easier workaround when the stereo imaging is important is to use Audacity's old Change Tempo effect (high quality)

Hm. Why being picky on stereo image and not the rest, after all? I'm starting to feel we're opening a can of worms here. I can heard examples where Staffpad's stretcher yielded better quality than SBSMS. @SteveDaulton while this workaround _may_ do a better job at preserving stereo image, it may do a worse one at e.g. transient preservation.
Maybe a better tip would simply be to make users aware that Audacity has several time-stretching alternatives to choose from and experiment with in case artefacts are noticed.

**Comment by Paul-Licameli** [comment]
A question about step 5 in the original post.

if you hard-pan the result left or right and listen, does it sound strange?  Can you describe it?

**Comment by saintmatthieu** [comment]
> Maybe a better tip would simply be to make users aware that Audacity has several time-stretching alternatives to choose from and experiment with in case artefacts are noticed.

... or event that dual-mono processing is also something worth trying if the observed artefact is one of stereo image degradation.

**Comment by petersampsonaudacity** [comment]
> ... or even that dual-mono processing is also something worth trying if the observed artefact is one of stereo image degradation.

**Which begs the question:**  why not get Audacity to do the dual-mono processing automatically?

**Comment by crsib** [comment]
> why not get Audacity to do the dual-mono processing automatically

Because generally, it works worse for the case where time stretching is needed.

**Comment by LWinterberg** [comment]
The issue of results such as what [Matt shared above](https://github.com/audacity/audacity/issues/5603#issuecomment-1808587202) for dual-mono are that when things are mixed down to mono, they cancel out perfectly. Given that mono is a major source of playbacks (especially in smartphones and portable radios), sacrificing a bit of the stereo image is preferable to annihilating transients on entire classes of systems.

**Comment by saintmatthieu** [comment]
I would agree with @crsib and @LWinterberg, but objectively, @petersampsonaudacity while we don't have data to support the claim that one of dual-mono or joint-stereo is generally better than the other, doing what you suggest may be counter-productive. In the TL;DR of https://github.com/audacity/audacity/issues/5603#issuecomment-1808587202 I suggested one possible enhancement request. Or maybe @Paul-Licameli and I could invest time in the StaffPad code to see if there are means to improve the algorithm itself. For obvious reasons I'd wait on overall user feedback before considering this, though.

**Comment by petersampsonaudacity** [comment]
> the generalization is in my opinion wrong. Why not "Time-stretching may affect stereo image" ?

@saintmatthieu   OK I just implemented that wording.

**Comment by crsib** [comment]
> In contrast if I take the original un-stretched clip of Sonny Rollins' Without A song and hard pan right I hear his sax loud and clear - if I pan hard left his sax is thin and weak.

I don't think that time-stretching the carefully mastered audio is a use case for real-time time-stretching. I don't understand why it is used as an argument. 

Floating point calculations are prone to errors. They can be reduced at the expense of performance but remain imperfect. For the purpose of this algorithm, mid/side yields the best result. Dual mono shows noticeable artifacts.

**Comment by Paul-Licameli** [comment]
I linked a draft PR.  It completely solves the problem with the toy example of a chirp and DTMF tones.

Try it out.  I don't know yet what quality compromise it will make with real world recordings.

**Comment by petersampsonaudacity** [comment]
Personally I actually don't really mind if the stereo image is degraded by Time-stretching clips (partly because I'm extremely unlikely to want or need use this functionality.

But I do strongly feel. that we need to be straightforwardly honest with users in the documentation as what the downside risks are when using time-stretching on clips _(which is what I have attempted to do in the Manual)_.

--------------------------------------------------------------------

BTW I've been experimenting other stretchy effects (on Sonny Rollins' Without a Song:
a) Change Speed (and Pitch)
b) Change Tempo - low quality (default)
c) Change Tempo - high quality
d) Paulstretch
e) Sliding Stretch

None of those changes the stereo image.

**Comment by crsib** [comment]
I have a strong opinion that in the discussion that "screwdriver is bad for hammering nails" we are trying to fix the screwdriver. 

Time-stretching the carefully mastered clips is not a goal. Proving a point for "toy" examples is not a point either. Assessments should be made carefully on samples/clips, not on how Sgt Pepper sounds.

---

Using a different algorithm during the export that yields significantly different results is not an option either.

**Comment by LWinterberg** [comment]
The fundamental issue this bug runs into would be that there's no way to preview the stretching effect. SBSMS is orders of magnitude too slow to use in a realtime capability, and if you actually massively care about which algorithm sounds best you'd constantly be exporting files to figure out whether you like the tweak better or not. 

There is an argument to be made to allow a choice between *real-time capable* algorithms in the future (rubberband comes to mind) in a clip properties dialog.

**Comment by Paul-Licameli** [comment]
I suggest this issue might be retitled and reopened as a bug fix.

**Comment by petersampsonaudacity** [comment]
> The build is done, have at it @petersampsonaudacity

@Paul-Licameli 

Yepper, that Sounds **_much_** better with a proper stereo image retained after the clip rendering.

It also looks to have a stereo image with differing left and right channels in the displayed waveform.  This is Sonny Rollins stretched to 110% after rendering:
![image](https://github.com/audacity/audacity/assets/50205984/84e21e53-0df3-4677-afff-298310da00b2)


Not the same as after stretch and before the render of course:
![image](https://github.com/audacity/audacity/assets/50205984/61e32042-0311-4327-a116-ae6f94cb04bf)


And for reference this is it un-stretched:
![image](https://github.com/audacity/audacity/assets/50205984/9f5b40c5-4799-4ee1-a2b4-78465aca92de)

**Comment by Paul-Licameli** [comment]
> > The build is done, have at it @petersampsonaudacity
> 
> @Paul-Licameli
> 
> Yepper, that Sounds **_much_** better with a proper stereo image retained after the clip rendering.
> 
> It also looks to have a stereo image with differing left and right channels in the displayed waveform. This is Sonny Rollins stretched to 110% after rendering: ![image](https://user-images.githubusercontent.com/50205984/284257297-84e21e53-0df3-4677-afff-298310da00b2.png)
> 
> Not the same as after stretch and before the render of course: ![image](https://user-images.githubusercontent.com/50205984/284257531-61e32042-0311-4327-a116-ae6f94cb04bf.png)
> 
> And for reference this is it un-stretched: ![image](https://user-images.githubusercontent.com/50205984/284257751-9f5b40c5-4799-4ee1-a2b4-78465aca92de.png)

The second picture has more dark blue stuff but about the same pale blue stuff (or slightly less).

The "slightly less" (some loss of RMS) is a problem we know about and @saintmatthieu left a draft somewhere of an experimental fix.

The more dark blue stuff is just because, "shift happens."  (Phase shift that is, of different frequency bins.)  We think it's just inherent and unavoidable that the waveform may have to change its shape.

**Comment by petersampsonaudacity** [comment]
> Have you shared your Beatles file so the rest of us can experiment with the same?

@Paul-Licameli 

I have sent you Sgt Pepper and Sonny Rollins  offlist in an email, so you can use those for testing yourself and looking at the spectrograms (if you still need to do so).

**Comment by petersampsonaudacity** [comment]
Fro drumbeats I experimented with Under My Thumb from The Rolling Stones Aftermath album with Charlie Watts hitting the skins (right in the intro).

They sound fine to me
a) after time-stretching
b) after rendering the time-stretch

_@Paul-Licameli  - I sent you a copy of that audio file by email_

**Comment by SteveDaulton** [comment]
> Why being picky on stereo image and not the rest, after all?

Because for small amounts of stretch, both "real-time" stretch and sbsms provide good overall quality, with hardly any audible difference in sound quality other than that one mangles to stereo field and the other one doesn't.

Consider the not-uncommon use case of needing to stretch a stereo track just a little to align with another track that has a slightly different tempo, or slightly stretching an audio track made with an audio recorder to replace an inferior audio track in a video of the same event (consumer level video cameras and audio recorders frequently run at slightly  different speeds due to variations in their clocks). In such cases, the mangling of the stereo field may be unacceptable to the user.

**Comment by Paul-Licameli** [comment]
> For drumbeats I experimented with Under My Thumb from The Rolling Stones Aftermath album with Charlie Watts hitting the skins (right in the intro).
> 
> They sound fine to me a) after time-stretching b) after rendering the time-stretch
> 
> _@Paul-Licameli - I sent you a copy of that audio file by email_

@LWinterberg however has shared one example of drums where my change causes one of the channels to invert phase for certain periods, and I don't know why.

So this easy fix I tried may have tradeoffs I don't understand yet.

**Comment by petersampsonaudacity** [comment]
I just tried with sandy Nelson's Let There Be Drums (perhaps the most definitive pop drum record - and cymbals too).

And that sounded fine to me  after a stretch to 110% and a render.

I'll send you a copy @Paul-Licameli

**Comment by petersampsonaudacity** [comment]
I also tried The Band Of H.M. Royal Marines playing Sussex By The Sea (my soccer team's theme tune - Brighton & Hove Albion).

Lots of drums on that too and sounds fine after a stretch to 110% and a render

**Comment by LWinterberg** [comment]
Stretching to 95% speed in paul's build has things inverted and out of phase at seemingly random points. I found it pretty easy to hear. When mixing down to mono, it results in extremely inconsistent peak heights, sometimes almost muting the output.

![image](https://github.com/audacity/audacity/assets/87814144/5aa2feda-d3a9-4af2-84f4-4f4615d5a4d2)

compare this to the current stable version, which behaves pretty well: 

![image](https://github.com/audacity/audacity/assets/87814144/44c1ce28-1782-4576-8a24-77d375f21ef5)

**Comment by Paul-Licameli** [comment]
Yes, mixing stereo to mono in the parts where only one channel flips phase would make destructive interference.

If you had a goniometer, or “jellyfish,” you would see those parts making a sideways oscillation instead of vertical.

Well not bad for less than two hours of my effort to hack the library and try an idea.

I have another crazy idea I might throw past @saintmatthieu

**Comment by Paul-Licameli** [comment]
Also in all these years I never dove into Soundtouch or SBSMS to figure out how they worked.  Maybe there ideas there.
